src/transport.rs

```diff
@@ -0,0 +1,473 @@
+        let (mut mixnet, inbound_rx, outbound_tx) = Mixnet::new(uri).await?;
+        let self_address = mixnet.get_self_address().await?;
+        let listen_addr = nym_address_to_multiaddress(self_address)?;
+        let listener_id = ListenerId::new();
+
+        #[allow(clippy::type_complexity)]
+        let (poll_tx, poll_rx): (
+            Sender<TransportEvent<Upgrade, NymTransportError>>,
+            Receiver<TransportEvent<Upgrade, NymTransportError>>,
+        ) = async_channel::unbounded();
+
+        poll_tx
+            .send(TransportEvent::NewAddress {
+                listener_id,
+                listen_addr: listen_addr.clone(),
+            })
+            .await?;
+
+        Ok(Self {
+            self_address,
+            listen_addr,
+            listener_id,
+            connections: HashMap::<ConnectionId, InnerConnection>::new(),
+            pending_dials: HashMap::<ConnectionId, PendingConnection>::new(),
+            inbound_rx,
+            outbound_tx,
+            poll_rx,
+            poll_tx,
+            mixnet,
+            waker: None,
+        })
+    }
+
+    // handle_connection_response resolves the pending connection corresponding to the response
+    // (if there is one) into a Connection.
+    fn handle_connection_response(&mut self, msg: ConnectionMessage) -> Result<(), Error> {
+        let pending_conn = self.pending_dials.get(&msg.id);
+        if pending_conn.is_none() {
+            return Err(anyhow!("no connection found for ConnectionRespone"));
+        }
+
+        if self.connections.contains_key(&msg.id) {
+            return Err(anyhow!(
+                "received ConnectionResponse but connection was already established"
+            ));
+        }
+
+        let pending_conn = pending_conn.unwrap();
+
+        // resolve connection and put into pending_conn channel
+        let conn = Connection::new(pending_conn.remote_recipient, msg.id.clone());
+        let inner_conn = InnerConnection::new(pending_conn.remote_recipient, msg.id.clone());
+        let connection_tx = pending_conn.connection_tx.clone();
+
+        tokio::task::spawn(async move {
+            connection_tx.send(conn).await.unwrap();
+        });
+
+        self.connections.insert(msg.id, inner_conn);
+        Ok(())
+    }
+
+    /// handle_connection_request handles an incoming connection request, sends back a
+    /// connection response, and finally completes the upgrade into a Connection.
+    fn handle_connection_request(&mut self, msg: ConnectionMessage) -> Result<Connection, Error> {
+        if msg.recipient.is_none() {
+            return Err(anyhow!(
+                "received None recipient in ConnectionRequest: {:?}",
+                msg
+            ));
+        }
+
+        // ensure we don't already have a conn with the same id
+        if self.connections.get(&msg.id).is_some() {
+            return Err(anyhow!(
+                "cannot handle connection request; already have connection with given ID"
```

Comment:
This should be a variant of NymTransportError, ie: `NymTransportError::ConnectionIdAlreadyInUsed(ConnectionId)`